### PR TITLE
Generate new cert if expiration date is too close.

### PR DESCRIPTION
**Pull request by pbalogh-sa**
fixes: #159

### CODE REVIEW

File: operator/pkg/controller/vault/vault_controller.go [hunk 1]
```diff
@@ -1685,3 +1749,22 @@ func (r *ReconcileVault) distributeCACertificate(v *vaultv1alpha1.Vault, caSecre
...
+	if block == nil {
+		return "", fmt.Errorf("failed to parse certificate PEM")
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse certificate: %v", err)
+	}
+
+	// If remaining time lower than one week return true
+	if cert.NotAfter.Sub(time.Now()).Hours() < 168 {
```
Comment (primeroz) [inline operator/pkg/controller/vault/vault_controller.go]: Should the expire threshold be configurable ?

Comment (pbalogh-sa) [reply operator/pkg/controller/vault/vault_controller.go]: You are right, I'm on it.

File: operator/pkg/controller/vault/vault_controller.go [hunk 2]
```diff
@@ -1175,6 +1219,26 @@ func withVaultWatchedExternalSecrets(v *vaultv1alpha1.Vault, secrets []corev1.Se
...
+	if sfsAnnotations["tlsExpiration"] != "" {
+		annotations["tlsExpiration"] = sfsAnnotations["tlsExpiration"]
+	}
+	if tlsExpiration != "" {
+		annotations["tlsExpiration"] = tlsExpiration
+	}
+	return annotations
+}
+
+func withTLSExpirationAnnotations(v *vaultv1alpha1.Vault, tlsAnnotations, annotations map[string]string) map[string]string {
```
Comment (baluchicken) [inline operator/pkg/controller/vault/vault_controller.go]:
```suggestion
func withTLSExpirationAnnotations(tlsAnnotations, annotations map[string]string) map[string]string {
```

File: operator/pkg/controller/vault/vault_controller.go [hunk 3]
```diff
@@ -1024,7 +1068,7 @@ func statefulSetForVault(v *vaultv1alpha1.Vault, externalSecretsToWatchItems []c
 			Template: corev1.PodTemplateSpec{
 				ObjectMeta: metav1.ObjectMeta{
 					Labels:      withVaultLabels(v, ls),
-					Annotations: withVaultAnnotations(v, withVaultWatchedExternalSecrets(v, externalSecretsToWatchItems, withPrometheusAnnotations("9102", getCommonAnnotations(v, map[string]string{})))),
+					Annotations: withTLSExpirationAnnotations(v, tlsAnnotations, withVaultAnnotations(v, withVaultWatchedExternalSecrets(v, externalSecretsToWatchItems, withPrometheusAnnotations("9102", getCommonAnnotations(v, map[string]string{}))))),
```
Comment (baluchicken) [inline operator/pkg/controller/vault/vault_controller.go]:
```suggestion
					Annotations: withTLSExpirationAnnotations(tlsAnnotations, withVaultAnnotations(v, withVaultWatchedExternalSecrets(v, externalSecretsToWatchItems, withPrometheusAnnotations("9102", getCommonAnnotations(v, map[string]string{}))))),
```

File: operator/pkg/controller/vault/vault_controller.go [hunk 4]
```diff
@@ -231,19 +233,48 @@ func (r *ReconcileVault) Reconcile(request reconcile.Request) (reconcile.Result,
...
 		}
 
 		// Set Vault instance as the owner and controller
 		if err := controllerutil.SetControllerReference(v, sec, r.scheme); err != nil {
 			return reconcile.Result{}, err
 		}
 
-		err = r.createObjectIfNotExists(sec)
+		err = r.createOrUpdateObject(sec)
+		//err = r.createObjectIfNotExists(sec)
```
Comment (baluchicken) [inline operator/pkg/controller/vault/vault_controller.go]:
```suggestion
```

File: operator/pkg/controller/vault/vault_controller.go [hunk 5]
```diff
@@ -1175,6 +1219,26 @@ func withVaultWatchedExternalSecrets(v *vaultv1alpha1.Vault, secrets []corev1.Se
 	return annotations
 }
 
+func getTLSExpirationAnnotations(statefulSet *appsv1.StatefulSet, tlsExpiration string) map[string]string {
+	annotations := map[string]string{}
+	sfsAnnotations := statefulSet.Spec.Template.GetAnnotations()
+	if sfsAnnotations["tlsExpiration"] != "" {
+		annotations["tlsExpiration"] = sfsAnnotations["tlsExpiration"]
```
Comment (primeroz) [inline operator/pkg/controller/vault/vault_controller.go]: should this annotation be "FQDN" name ? 

similar to https://github.com/banzaicloud/bank-vaults/blob/4fe73a21c6ce30052861f59be446928406d03e4a/operator/pkg/controller/vault/vault_controller.go#L1217

Comment (pbalogh-sa) [reply operator/pkg/controller/vault/vault_controller.go]: nice catch

File: operator/pkg/controller/vault/vault_controller.go [hunk 6]
```diff
@@ -231,19 +235,48 @@ func (r *ReconcileVault) Reconcile(request reconcile.Request) (reconcile.Result,
 		}
 	}
 
+	tlsExpiration := time.Time{}
 	if !v.Spec.GetTLSDisable() {
 		// Create the secret if it doesn't exist
```
Comment (primeroz) [inline operator/pkg/controller/vault/vault_controller.go]: Bit picky i know, but i guess this comment does not apply anymore

File: operator/cmd/manager/main.go [hunk 7]
```diff
@@ -65,6 +65,7 @@ func handleLiveness() {
 func main() {
 
 	syncPeriod := flag.Duration("sync_period", 30*time.Second, "SyncPeriod determines the minimum frequency at which watched resources are reconciled")
+	tlsExpTH := flag.Float64("tls_exp_th", 168, "TLS cert expiration treshold in hours")
```
Comment (pepov) [inline operator/cmd/manager/main.go]: wouldn't this be better configured in the CR?

Comment (bonifaido) [reply operator/cmd/manager/main.go]: I was thinking the same!

Comment (pbalogh-sa) [reply operator/cmd/manager/main.go]: I agree with you. It will be better in CR.

**Pull request merged**